linkerd/lock/src/test.rs

```diff
@@ -0,0 +1,254 @@
+            .map(move |_| {
+                // svc1 grabs the lock and fails immediately.
+                let mut svc1 = svc0.clone();
+                assert!(svc1
+                    .poll_ready()
+                    .expect_err("mut fail")
+                    .downcast_ref::<ServiceError>()
+                    .expect("must fail with service error")
+                    .inner()
+                    .is::<Underflow>());
+
+                // svc0 suffers the same fate.
+                assert!(svc0
+                    .poll_ready()
+                    .expect_err("mut fail")
+                    .downcast_ref::<ServiceError>()
+                    .expect("must fail with service error")
+                    .inner()
+                    .is::<Underflow>());
+            })
+    }));
+}
+
+#[test]
+fn dropping_releases_access() {
+    use tower::util::ServiceExt;
+    run(future::lazy(|| {
+        let mut svc0 = Lock::new(Decr::new(3, Arc::new(true.into())));
+
+        // svc0 grabs the lock, but the inner service isn't ready.
+        assert!(svc0.poll_ready().expect("must not fail").is_ready());
+
+        let svc1 = svc0.clone();
+        let (tx1, rx1) = oneshot::channel();
+        tokio::spawn(
+            svc1.oneshot(1)
+                .then(move |_| {
+                    trace!("complete");
+                    tx1.send(()).map_err(|_| ())
+                })
+                .instrument(info_span!("Svc1")),
+        );
+
+        let svc2 = svc0.clone();
+        let (tx2, rx2) = oneshot::channel();
+        tokio::spawn(
+            svc2.oneshot(1)
+                .then(move |_| {
+                    trace!("complete");
+                    tx2.send(()).map_err(|_| ())
+                })
+                .instrument(info_span!("Svc2")),
+        );
+
+        // svc3 will be the notified waiter when svc0 completes; but it drops
+        // svc3 before polling the waiter. This test ensures that svc2 is
+        // notified by svc3's drop.
+        let svc3 = svc0.clone();
+        let (tx3, rx3) = oneshot::channel();
+        tokio::spawn(PollAndDrop(svc3, rx3).instrument(info_span!("Svc3")));
+
+        tokio::spawn(
+            svc0.ready()
+                .then(move |_| tx3.send(()).map_err(|_| ()))
+                .instrument(info_span!("Svc0")),
+        );
+        // svc3 notified; but it is dropped before it can be polled
+
+        rx2.then(move |_| rx1).map_err(|_| ())
+    }));
+
+    struct PollAndDrop(Lock<Decr>, oneshot::Receiver<()>);
+    impl Future for PollAndDrop {
+        type Item = ();
+        type Error = ();
+        fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+            trace!("Polling");
+            if self.1.poll().map_err(|_| ())?.is_ready() {
+                trace!("Dropping");
+                return Ok(Async::Ready(()));
+            }
+            self.0.poll_ready().map_err(|_| ())
+        }
+    }
+}
+
+#[test]
+fn fuzz() {
```

Comment:
You may want to consider using `loom` for this test. `loom` 0.1.x has [a mock version of `AtomicTask`](https://github.com/tokio-rs/loom/blob/2f735ffc5485edd98cb8158212dcf89bad1ad98b/src/futures/atomic_task.rs#L8-L15), which can simulate spurious wakeups, and a mock version of `std::thread`, which will run every permutation of threads being preempted (with bounding).

Using `loom` rather than randomness would make this task deterministic, which could make debugging failures significantly easier.

Reply:
Sounds cool! I'd rather not block the PR on this unless we have specific concerns.